app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java

```diff
@@ -1908,13 +1914,24 @@ private void removeMarker(Place place){
 
     @Override
     public void recenterMap(fr.free.nrw.commons.location.LatLng currentLatLng) {
-        if (isPermissionDenied || currentLatLng == null) {
-            recenterToUserLocation = true;
-            checkPermissionsAndPerformAction();
-            if (!isPermissionDenied && !(locationManager.isNetworkProviderEnabled()
-                || locationManager.isGPSProviderEnabled())) {
-                showLocationOffDialog();
+        // if user has denied permission twice, then show dialog
+        if (isPermissionDenied) {
+            if (locationPermissionsHelper.checkLocationPermission(getActivity())) {
+                // this will run when user has given permission by opening app's settings
+                isPermissionDenied = false;
+                recenterMap(currentLatLng);
```

Comment:
Is there a specific reason this is a recursive call? Wouldn't a direct call to the `locationPermissionGranted()` method be sufficient here? 

With recursion, I suppose we run the risk of invoking the `addCurrentLocationMarker(currentLatLng)` method and code that follows it twice. As a consequence, the map marker could be added twice.

Reply:
> Is there a specific reason this is a recursive call? Wouldn't a direct call to the `locationPermissionGranted()` method be sufficient here?

Actually, `recenterMap()` is called when the user clicks on the target button, and `locationPermissionGranted()` doesn't serve that purpose, so I added a recursive call here, otherwise I had click on the target button twice to recenter the marker.

> With recursion, I suppose we run the risk of invoking the `addCurrentLocationMarker(currentLatLng)` method and code that follows it twice. As a consequence, the map marker could be added twice.

Um, the map marker won't be added twice as `addCurrentLocationMarker(currentLatLng)` only calls `recenterMarkerToPosition()`, which recenters the marker, not add another one.

Reply:
> Actually, recenterMap() is called when the user clicks on the target button, and locationPermissionGranted() doesn't serve that purpose, so I added a recursive call here, otherwise I had click on the target button twice to recenter the marker.

I don't get this part. Specifically, I'm not sure why a recrusive call is necessary. Let me elaborate. With you change, we at some point enter the `recenterMap` function. Let us assume `isPermissionDenied` is `true`. So, we enter the first `if` branch and check if the app actually has location permission or not. If it does (as the user could've granted it via the "Settings" screen), we call `recenterMap` recursively. Now, `isPermissionDenied` would be `false` because we explicitly set that in the `if` branch. So, we get to the first `else` branch. There we could easily presume `locationPermissionsHelper.checkLocationPermission` would return `true` since if just returned so in the previous call. So, the nested `else` branch is taken. Here the `locationPermissionGranted` method is called and the flow proceeds after that to the following `if (currentLatLng == null)` check and `addCurrentLocationMarker` method calls.

Here's the interesting part about the recursion. After the whole execution of the `recenterMap` method, the method returns and we're now back again in the first if of the `recenterMap` method (the invoker of the recursive call). The flow continues to the following `if (currentLatLng == null)` check and `addCurrentLocationMarker` method calls. Notice how this is being invoked for a second time.

This second invocation is precisely what I'm not sure is warranted. Instead, if we replace the recursive call with an invocation to the `locationPermissionGranted` method (which seems to be the only additional thing being done via the recursive call), then we get to have the whole flow without any recursion. Is this not the case?

Let me know in case I missed something.

> Um, the map marker won't be added twice as addCurrentLocationMarker(currentLatLng) only calls recenterMarkerToPosition(), which recenters the marker, not add another one.

Ah ok. Got it :+1:

Reply:
I understand your apprehensions regarding the recursion call and to remove that, I can add a return statement after the call, so that the method runs fully only once. But, I think we will have to call recenterMap() only and not `locationPermissionGranted()` as`recenterMap()` makes the `mapView.getController().animateTo(new GeoPoint(currentLatLng.getLatitude(), currentLatLng.getLongitude()));` which is not done by `locationPermissionGranted()`. Thus the map will not animate to current location when target button is clicked (if we call `locationPermissionGranted()` instead of `recenterMap()`

Reply:
> I understand your apprehensions regarding the recursion call and to remove that, I can add a return statement after the call, so that the method runs fully only once.

I could understand the `return` after the recursive call would do the trick. The thing is the recursive call seems totally unwarranted there. I'll elaborate a bit below.

> But, I think we will have to call recenterMap() only and not `locationPermissionGranted()` as`recenterMap()` makes the `mapView.getController().animateTo(new GeoPoint(currentLatLng.getLatitude(), currentLatLng.getLongitude()));` which is not done by `locationPermissionGranted()`. Thus the map will not animate to current location when target button is clicked (if we call `locationPermissionGranted()` instead of `recenterMap()`

I think there's some confusion on the change I'm suggesting. Here's a diff showcasing the specific change that I suggest:

```diff
             if (locationPermissionsHelper.checkLocationPermission(getActivity())) {
                 // this will run when user has given permission by opening app's settings
                 isPermissionDenied = false;
-                recenterMap(currentLatLng);
+                locationPermissionGranted();
             } else {
                 askForLocationPermission();
             }
```

With this change, the call to `mapView.getController().animateTo()` would still happen (if we need to be precise, the call to `binding.map.getController().animateTo` owing to recent changes). That's because the flow would still continue after the `locationPermissionGranted` method is done. Feel free to test and confirm this.

Reply:
Sure, I'll check this once and get back to you.

Reply:
Were you able to check this @ShashwatKedia ?

This is the only pending thread in this MR, so far. If you could check and handle it, we could proceed with merging it.